File: Sources/Runtime/Components/MelodyStepper.swift
import SwiftUI
import Core

#if os(tvOS)
#else
/// Renders a stepper control bound to a screen state key with configurable range and step.
struct MelodyStepper: View {
    let definition: ComponentDefinition
    let resolvedLabel: String
    var onChanged: (() -> Void)? = nil

    @Environment(\.screenState) private var screenState

    var body: some View {
        Stepper(
            resolvedLabel,
            value: binding,
            in: minValue...maxValue,
            step: stepValue
        )
        .melodyStyle(definition.style)
        .onChange(of: binding.wrappedValue) { _, _ in
            onChanged?()
        }
    }

    private var binding: Binding<Double> {
        guard let key = definition.stateKey else {
            return .constant(0)
        }
        return Binding(
            get: { screenState.get(key: key).numberValue ?? 0 },
            set: { screenState.set(key: key, value: .number($0)) }
        )
    }

    private var minValue: Double { definition.min ?? 0 }
    private var maxValue: Double { definition.max ?? 100 }
    private var stepValue: Double { definition.step ?? 1 }
}
#endif

File: Android/melody-runtime/src/main/java/com/melody/runtime/components/MelodyStepper.kt
package com.melody.runtime.components

import androidx.compose.foundation.layout.*
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Add
import androidx.compose.material.icons.filled.Remove
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.unit.dp
import com.melody.core.schema.ComponentDefinition
import com.melody.runtime.engine.LuaValue
import com.melody.runtime.renderer.LocalIsDisabled
import com.melody.runtime.renderer.LocalScreenState

@Composable
fun MelodyStepper(
    definition: ComponentDefinition,
    resolvedLabel: String,
    onChanged: (() -> Unit)? = null
) {
    val screenState = LocalScreenState.current
    val isDisabled = LocalIsDisabled.current
    val stateKey = definition.stateKey
    val minValue = definition.min ?: 0.0
    val maxValue = definition.max ?: 100.0
    val stepValue = definition.step ?: 1.0

    val currentValue = if (stateKey != null) {
        screenState.slot(stateKey).value.numberValue ?: 0.0
    } else {
        0.0
    }

    Row(
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(8.dp),
        modifier = Modifier.melodyStyle(definition.style)
    ) {
        if (resolvedLabel.isNotEmpty()) {
            Text(resolvedLabel, modifier = Modifier.weight(1f))
        }

        IconButton(
            onClick = {
                val newValue = (currentValue - stepValue).coerceAtLeast(minValue)
                if (stateKey != null) {
                    screenState.set(stateKey, LuaValue.NumberVal(newValue))
                }
                onChanged?.invoke()
            },
            enabled = !isDisabled && currentValue > minValue
        ) {
            Icon(Icons.Default.Remove, contentDescription = "Decrease")
        }

        Text(
            text = if (currentValue == currentValue.toLong().toDouble()) {
                currentValue.toLong().toString()
            } else {
                currentValue.toString()
            }
        )

        IconButton(
            onClick = {
                val newValue = (currentValue + stepValue).coerceAtMost(maxValue)
                if (stateKey != null) {
                    screenState.set(stateKey, LuaValue.NumberVal(newValue))
                }
                onChanged?.invoke()
            },
            enabled = !isDisabled && currentValue < maxValue
        ) {
            Icon(Icons.Default.Add, contentDescription = "Increase")
        }
    }
}
